### PR TITLE
Menu button widget

### LINKED ISSUE #3261: [Feature] Add Menu Widget
## Summary

Add 3 dot button that presents a menu of action options

## Motivation

Often in constrained spaced UIs it's necessary to provide a set of actions but there is not enough space to have separate buttons for each action. In these situations a 3 dot button is needed that when selected presents a set of available action.

## Additional Context

The 3 dot button is a variation of the [hamburger button](https://en.wikipedia.org/wiki/Hamburger_button) but is used in places other than just the top corner of the UI to reveal the main menu.

This 3 dot button can come in both vertical (aka "kebab") and horizontal (aka "meatballs") layout as per [StackOverflow: What is the hamburger menu icon called and the three vertical dots icon called?](https://stackoverflow.com/questions/34755666/what-is-the-hamburger-menu-icon-called-and-the-three-vertical-dots-icon-called)

The visual aspect of the 3 dot menu should be configurable.

# Developer Handoff Design
Link to the developer Handoff Document: [Figma link](https://www.figma.com/file/tVBQrt0G0XI89eS9m8cT5H/Menu-Widget?node-id=1%3A27)

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description

- First MVP of the whole widget

Fixes #3261 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/3261-menu-widget 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 53.75 **(-0.1)** | 35.77 **(-0.04)** | 32.3 **(-0.19)** | 54.32 **(-0.11)**
 :sparkles: :new: | **app/client/src/components/designSystems/appsmith/MenuButtonComponent/index.tsx** | **23.81** | **28.57** | **0** | **23.81**
 :sparkles: :new: | **app/client/src/components/propertyControls/IconAlignControl.tsx** | **60.87** | **40** | **25** | **61.9**
 :sparkles: :new: | **app/client/src/components/propertyControls/IconSelectControl.tsx** | **41.67** | **23.53** | **20** | **43.64**
 :sparkles: :new: | **app/client/src/components/propertyControls/MenuItemsControl.tsx** | **27.63** | **40** | **9.09** | **27.78**
 :green_circle: | app/client/src/components/propertyControls/index.ts | 83.87 **(1.73)** | 100 **(0)** | 0 **(0)** | 83.33 **(1.85)**
 :red_circle: | app/client/src/icons/ControlIcons.tsx | 57.58 **(-0.31)** | 100 **(0)** | 12.5 **(-0.54)** | 57.58 **(-0.31)**
 :red_circle: | app/client/src/icons/WidgetIcons.tsx | 64.29 **(-0.52)** | 100 **(0)** | 25.93 **(-0.99)** | 64.29 **(-0.52)**
 :red_circle: | app/client/src/utils/WidgetRegistry.tsx | 74.75 **(-0.25)** | 100 **(0)** | 24.24 **(-0.76)** | 74.75 **(-0.25)**
 :sparkles: :new: | **app/client/src/widgets/MenuButtonWidget.tsx** | **69.57** | **0** | **33.33** | **71.43**</details>

### CODE REVIEW

**Review by somangshu — APPROVED** [review]
Approving to run  tests and rebuild deploy preview

**Comment by somangshu** [comment]
@wmdev0808 is the feature complete, I am see differences from what was designed.

**Review by somangshu — DISMISSED** [review]
Approving to rebuild deploy preview

File: app/client/src/components/propertyControls/IconSelectControl.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            width: 100%;
+            text-align: center;
+          }
+        `}
+        icon={icon === NONE ? undefined : icon}
+        key={icon}
+        onClick={handleClick}
+        text={icon}
+      />
+      // <Icon
```
Comment (mohanarpit) [inline app/client/src/components/propertyControls/IconSelectControl.tsx:99]: Please remove all commented code. Ensures that the code stays clean.

**Comment by somangshu** [comment]
Collecting feedback: **WIP** 🚧 

- The icon control design needs to be grid and follow the design initially given
- The widget icon has a hover state.
- Buttons inside the menu should not have a padding on the left and right
- The compact mode should squeeze menu options rather than the component itself
- The background on the Icon control component needs to be `#ffffff`
- The icon alignment control needs to follow design

@jsartisan we need clarification on this
- The menu button has some shadow which is looking weird. Can we stick to the default style similar to our button widget?
- The control for icon alignments is a little big. The height should be similar to other controls ( sorry in case you follow the height from the design )

**Comment by hetunandu** [comment]
@wmdev0808 The icon list now looks like this. Is this intended? Was it supposed to be a grid?

screenshot


<details><summary>Browser metadata</summary>





```
Path:      /applications/60d9c7414410c67f27c99348/pages/60d9c7414410c67f27c9934a/edit
Browser:   Firefox 89.0 on Mac OS 10.15
Viewport:  1680 x 846 @2x
Language:  en-US
Cookies:   Enabled
```

[View on BrowserStack](https://live.browserstack.com/dashboard#url=https%3A%2F%2Fdeploy-preview-5334--appsmith-editor-ce.netlify.app%2Fapplications%2F60d9c7414410c67f27c99348%2Fpages%2F60d9c7414410c67f27c9934a%2Fedit&os=OS+X&os_version=Catalina&browser=Firefox&browser_version=89.0&resolution=1680x846&speed=1&start=true)



</details>

[Open Deploy Preview](https://deploy-preview-5334--appsmith-editor-ce.netlify.app/applications/60d9c7414410c67f27c99348/pages/60d9c7414410c67f27c9934a/edit) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60daf45409818a000860453a&commentID=60daf9c15bc01ba028ec637b&resolution=resolved)

**Comment by wmdev0808** [comment]
> @wmdev0808 The icon list now looks like this. Is this intended? Was it supposed to be a grid?
> 
> screenshot
> 
> Browser metadata
> ```
> Path:      /applications/60d9c7414410c67f27c99348/pages/60d9c7414410c67f27c9934a/edit
> Browser:   Firefox 89.0 on Mac OS 10.15
> Viewport:  1680 x 846 @2x
> Language:  en-US
> Cookies:   Enabled
> ```
> 
> [View on BrowserStack](https://live.browserstack.com/dashboard#url=https%3A%2F%2Fdeploy-preview-5334--appsmith-editor-ce.netlify.app%2Fapplications%2F60d9c7414410c67f27c99348%2Fpages%2F60d9c7414410c67f27c9934a%2Fedit&os=OS+X&os_version=Catalina&browser=Firefox&browser_version=89.0&resolution=1680x846&speed=1&start=true)
> 
> [Open Deploy Preview](https://deploy-preview-5334--appsmith-editor-ce.netlify.app/applications/60d9c7414410c67f27c99348/pages/60d9c7414410c67f27c9934a/edit) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60daf45409818a000860453a&commentID=60daf9c15bc01ba028ec637b&resolution=resolved)

@hetunandu it should be grid

**Comment by wmdev0808** [comment]
> @wmdev0808 The icon list now looks like this. Is this intended? Was it supposed to be a grid?
> 
> screenshot
> 
> Browser metadata
> [Open Deploy Preview](https://deploy-preview-5334--appsmith-editor-ce.netlify.app/applications/60d9c7414410c67f27c99348/pages/60d9c7414410c67f27c9934a/edit) · [Mark as Resolved](https://app.netlify.com/cdp/resolve?deployID=60daf45409818a000860453a&commentID=60daf9c15bc01ba028ec637b&resolution=resolved)

@hetunandu It should be 3-columns grid

**Comment by somangshu** [comment]
> @hetunandu It should be 3-columns grid

@wmdev0808 I see it correctly. 

Feedback on this:
- Remove the names from the grid, Add 4 columns.
- Follow the design style as it is, The selected and hover color.

**Comment by wmdev0808** [comment]
> > @hetunandu It should be 3-columns grid
> 
> @wmdev0808 I see it correctly.
> 
> Feedback on this:
> 
> * Remove the names from the grid, Add 4 columns.
> * Follow the design style as it is, The selected and hover color.

**Comment by jsartisan** [comment]
Feedback:

1. The icon in the alignment control is not properly centered.
2. There is some border radius assigned to the alignment control button. We follow sharp edges in all our buttons. So border-radius:0 would be appropriate.
3.  The dropdown in icon control has some padding and width issues. Can we make the width of the dropdown the same as the control?
4. The search box is looking odd ( mainly because it is sticked to the edges ). There should be some margin like we had in design.
5. The scrollbar in the dropdown of icon control is ruining the whole control ( i am on windows and it looks really ugly ). We can style the scrollbar and make it look good.
Screenshot_92
6. The empty state is not right.
Screenshot_94
7. The icon control also has some border radius. We should follow sharp edges.

**Comment by somangshu** [comment]
@wmdev0808 We also need to add new documentation here.

**Comment by areyabhishek** [comment]
@wmdev0808 @jsartisan Can the width of the dropdown be the same as the width of the dropdown field? It looks weird to have dropdowns that are smaller in width. Another example below. 
Screen Shot 2021-07-01 at 2 12 48 AM

**Comment by somangshu** [comment]
@wmdev0808 few changes from the above comments are still pending.

Additional issues:
- Changes in the helper texts:
    - Compact mode: Decides if menu items will consume lesser space
    - Icon: Sets the icon to be used in the menu button.

**Review by somangshu — APPROVED** [review]
Approving to run tests

@jsartisan kindly code review

**Comment by YogeshJayaseelan** [comment]
@wmdev0808 @somangshu We have started testing this feature and came up with the below observations and still testing is in-progress 

1) When the disable button is enabled - Grey out is not happening and on hovering on the add menu widget - Not selectable icon is not displayed
https://loom.com/share/0250668bab1c414fb1193327f21c463a

2) Drop down size is smaller than the widget size 
Screenshot 2021-07-08 at 11 35 12 AM

3) If the label is removed left and right alignments are not working for icons 
https://loom.com/share/9ed23d0829e94667868b8fda314a9031

4) The 'Menu Button' text truncates when placing 'Add Menu Button' widget inside container/form/tabs/list widgets 
https://loom.com/share/15402439065a4a298836e18b1563d7ca

**Comment by wmdev0808** [comment]
> @wmdev0808 @somangshu We have started testing this feature and came up with the below observations and still testing is in-progress
> 
> 1. When `disable` is enabled - The button should grey our and use the `not allowed` cursor
> 
> Screenshot 2021-07-08 at 1 25 02 PM
> 
> 1. ~Drop down size is smaller than the widget size~ the menu width remain fixed regardless of the widget size
> 2. ~If the label is removed left and right alignments are not working for icons~ Alignment is not required with a single centered item
>    https://loom.com/share/9ed23d0829e94667868b8fda314a9031
> 3. The 'Menu Button' text truncates when placing 'Add Menu Button' widget inside container/form/tabs/list widgets
>    https://loom.com/share/15402439065a4a298836e18b1563d7ca
> 4. If the Menu names are longer then the text hovers with settings and Delete button
>    https://loom.com/share/25f67d6ee15a484890ed2a41f6d5fd4
> 
> 6)Enhancement - Icon names can be displayed as tool tip message on hovering on icons
> https://loom.com/share/4ce5706af432437bb324f8d3550d87db
> 
> 7)On the filter icon search if the search result has less options then the modal moves down and ui looks weird - This is happening in 16-inch (3072 × 1920)
> Screenshot 2021-07-08 at 2 55 07 PM



> @wmdev0808 @somangshu We have started testing this feature and came up with the below observations and still testing is in-progress
> 
> 1. When `disable` is enabled - The button should grey our and use the `not allowed` cursor
> 
> Screenshot 2021-07-08 at 1 25 02 PM
> 
> 1. ~Drop down size is smaller than the widget size~ the menu width remain fixed regardless of the widget size
> 2. ~If the label is removed left and right alignments are not working for icons~ Alignment is not required with a single centered item
>    https://loom.com/share/9ed23d0829e94667868b8fda314a9031
> 3. The 'Menu Button' text truncates when placing 'Add Menu Button' widget inside container/form/tabs/list widgets
>    https://loom.com/share/15402439065a4a298836e18b1563d7ca
> 4. If the Menu names are longer then the text hovers with settings and Delete button
>    https://loom.com/share/25f67d6ee15a484890ed2a41f6d5fd43
@somangshu, this is not the issue specific to this widget implementation, this issue comes from WigetNameComponent
> 6)Enhancement - Icon names can be displayed as tool tip message on hovering on icons
> https://loom.com/share/4ce5706af432437bb324f8d3550d87db
> 
> 7)On the filter icon search if the search result has less options then the modal moves down and ui looks weird - This is happening in 16-inch (3072 × 1920)
> Screenshot 2021-07-08 at 2 55 07 PM



> @wmdev0808 @somangshu We have started testing this feature and came up with the below observations and still testing is in-progress
> 
> 1. When `disable` is enabled - The button should grey our and use the `not allowed` cursor
> 
> Screenshot 2021-07-08 at 1 25 02 PM
> 
> 1. ~Drop down size is smaller than the widget size~ the menu width remain fixed regardless of the widget size
> 2. ~If the label is removed left and right alignments are not working for icons~ Alignment is not required with a single centered item
>    https://loom.com/share/9ed23d0829e94667868b8fda314a9031
> 3. The 'Menu Button' text truncates when placing 'Add Menu Button' widget inside container/form/tabs/list widgets
>    https://loom.com/share/15402439065a4a298836e18b1563d7ca
> 4. If the Menu names are longer then the text hovers with settings and Delete button
>    https://loom.com/share/25f67d6ee15a484890ed2a41f6d5fd43
> 
> 6)Enhancement - Icon names can be displayed as tool tip message on hovering on icons
> https://loom.com/share/4ce5706af432437bb324f8d3550d87db
> 
> 7)On the filter icon search if the search result has less options then the modal moves down and ui looks weird - This is happening in 16-inch (3072 × 1920)
> Screenshot 2021-07-08 at 2 55 07 PM

With regard to 4. @somangshu, this is not the issue specific to this widget implementation, this issue comes from WigetNameComponent

**Comment by wmdev0808** [comment]
@somangshu ,
With regard to 4, it is not the issue specific to this widget implementation, it just comes from WidgetNameComponent itself

File: app/client/src/components/designSystems/appsmith/MenuButtonComponent/index.tsx [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+import { ComponentProps } from "components/designSystems/appsmith/BaseComponent";
+import { darkenActive, darkenHover } from "constants/DefaultTheme";
+
+export const MenuButtonContainer = styled.div`
+  width: 100%;
+  height: 100%;
+  text-align: center;
+
+  & > .bp3-popover2-target {
```
Comment (jsartisan) [inline app/client/src/components/designSystems/appsmith/MenuButtonComponent/index.tsx:15]: Can we use .${Classes.POPOVER_TARGET} for the selector?

you can import it from blueprint core.
 import { Classes } from "@blueprintjs/core";

**Review by jsartisan — DISMISSED** [review]

File: app/client/src/components/designSystems/appsmith/MenuButtonComponent/index.tsx (hunk 2 above)
Comment (wmdev0808) [reply app/client/src/components/designSystems/appsmith/MenuButtonComponent/index.tsx]: @jsartisan, I've made a new commit for this comment

**Review by somangshu — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@wmdev0808 @somangshu Retested this feature and please find the status below,

**Issues fixed:**
1,2,3,5

**Issues needs fix:**
4) The 'Menu Button' text truncates when placing 'Add Menu Button' widget inside container/form/tabs/list widgets 

6) Sixth issue is fixed - But there is a problem on mouse hovering on any icons and scrolling up the tool tips goes out of icon property and by scrolling down the canvas flickers
https://loom.com/share/dee5abd1c76243faa52e3f5db551319c

**Comment by somangshu** [comment]
@YogeshJayaseelan 
- Point 4: The issue is related to widget tag and needs to be logged separately; We need to test this throughly and point most scenarios.
- Point 6: The issue is related to popper and looks like the default behaviour. Please log this issue separately if required.

**Comment by YogeshJayaseelan** [comment]
@somangshu logged issues for point 4 and 6 separately and the rest of the scenarios are working as expected

**Comment by somangshu** [comment]
Reran jobs again, At first the just build was failing, The second time `Execution_param.spec.js` failed. 

cc @NandanAnantharamu, @yatinappsmith

**Pull request merged**